apps/emqx_utils/src/emqx_utils.erl

```diff
@@ -999,6 +1000,36 @@ search(ExpectValue, KeyFunc, [Item | List]) ->
         false -> search(ExpectValue, KeyFunc, List)
     end.
 
+%% @doc Maps over a term or a list of terms and flattens the result, giving back
+%% again a term or a flat list of terms. It's similar to `lists:flatmap/2`, but
+%% it works on a single term as well, both as input and `Fun` output (thus, the
+%% wordplay on "flatter").
+%% The purpose of this function is to adapt to `Fun`s that return either a `[]`
+%% or a term, and to avoid costs of list construction and flattening when dealing
+%% with large lists.
+-spec flattermap(Fun, FlatList) -> FlatList when
+    Fun :: fun((X) -> FlatList),
+    FlatList :: [X] | X.
+flattermap(_Fun, []) ->
+    [];
+flattermap(Fun, [X | Xs]) ->
+    flatcomb(Fun(X), flattermap(Fun, Xs));
+flattermap(Fun, X) ->
+    Fun(X).
```

Comment:
Q: Shouldn't it return a `[Fun(X)]` to avoid "dependent typing"? I think it'll be better for composability, so when we write `F(flattermap(G, Arg))` F could always assume the result is a list?